### DISCUSSION TITLE
Thoughts on state management

### DISCUSSION THREAD

**Discussion by sloan58** [livewire/livewire#4774] [2. Show and tell]
I'm working on a project and wanted to have some mechanism to track user state across pages with Livewire.  The state that I'm tracking isn't mission critical.  If things get cleared, it's not the end of the world.  I'm currently using this in my Users Table component with search and paginating properties.  Just wondering if I'm on the right track here.  Am I reinventing the wheel?  Any glaring "gotchas" or inefficient code style that I might be missing.  Any feedback is appreciated.

**ManagesState Trait**

```php

<?php

namespace App\Http\Livewire\Traits;

use Illuminate\Support\Facades\Route;
use Illuminate\Support\Facades\Cache;

trait ManagesState
{
    public array $state;
    public string $cacheKey;
    public string $routeName;

    public function mountManagesState()
    {
        $this->cacheKey = sprintf(
            '%s.state',
            auth()->user()->id,
        );
        $this->routeName = Route::currentRouteName();
        $this->state = Cache::get($this->cacheKey)[$this->routeName] ?? [];
    }

    public function bootedManagesState() {
        foreach($this->statefulProps as $prop) {
            $this->{$prop} = $this->state[$prop] ?? $this->{$prop} ?? '';
        }
    }

    public function updatedManagesState($prop)
    {
        if(in_array($prop, $this->statefulProps)) {
            $this->setState([
                $prop => $this->{$prop}
            ]);
        }
    }

    public function setState($update)
    {
        $mutatingState = Cache::get($this->cacheKey) ?? [];
        $mutatingState[$this->routeName] = array_merge(
            $mutatingState[$this->routeName] ?? [],
            $update
        );
        Cache::forever($this->cacheKey, $mutatingState);
    }
}
```

**Users Table**

```php

<?php

namespace App\Http\Livewire\User;

use Livewire\Component;
use Livewire\WithPagination;
use App\Http\Livewire\Traits\ManagesState;

class Index extends Component
{
    use WithPagination, ManagesState;

    public string $filter;
    public string $paginating = '10';
    public array $statefulProps = ['filter', 'paginating'];

    ....
}

```

By defining the `$statefulProps` property in the components that use the ManagesState trait, it will register those properties to track them in the cache between page views, storing them in a unique array inside the user state cache key, identified by the route name so there's no collision with overlapping field names (like `$filter` and `$paginating`).  If the property has a default value in the component it will use that unless that key exists in the cache.

It seems to be working fine so far.  Is there anything that could/should be improved?  Thanks in advance.

**Comment by joshhanley** [comment]
@sloan58 looks all ok to me and if it works, then happy days!

What sort of things are you using this for?

Only thing I would do different, is I wouldn't store everything under one state key, I'd use a key per route by adding the route name to your current key. That then simplifies the retrieval and storage logic.

**Reply by sloan58** [reply]
Hey, @joshhanley!  Thank you for taking the time to check it out and provide feedback.  Much appreciated. I plan on using this to remember stuff like strings for search inputs, pagination settings, tab selections, and anything that helps to enhance the UX but that wouldn't be shocking if it just went away all of a sudden.  I may expand it out to track more data between page views, kinda like a front-end redux or something similar, but nothing that needs to be persisted long term.

I think you're right about the cache key as well.  I actually started out that way and when I was looking at it in the redis cli I thought it might get "cluttered", since it would be a unique key per user and per route, but who in the heck is watching the redis cache keys anyway?  I was sacrificing clean code in PHP for that - not worth it!  Thanks for the suggestion.  I'll def make that update.

Cheers
Marty

**Reply by joshhanley** [reply]
@sloan58 no worries! 🙂

So one thing I'm not sure if you're aware but you can use Livewire's querystring to keep track of a lot of that stuff and it gets remembered in history state, so if I a user goes back to that page (use browser back button), all the filters and stuff would be restored. That doesn't help if they use navigation to return to the page of course.

Yeah I'd opt for cleaner code than a clean redis cli (I don't think I've ever looked at it 😆).
